### ISSUE TITLE
Allow custom vfs implementations

### ISSUE THREAD

**Issue by rklaehn** [rusqlite/rusqlite#1145]
It would be nice if rusqlite had an **official** way to hoop up a custom VFS. This is frequently needed when using sqlite in "unusual" environments, such as wasm.

There is a crate that allows specifying a vfs: https://github.com/rkusa/sqlite-vfs . I am currently trying to get sqlite to work in the browser with persistence (not just memory), and I have adapted it to my own needs here: https://github.com/rklaehn/sqlite-vfs/tree/wasm

In https://github.com/rusqlite/rusqlite/pull/1010, there is also a new vfs impl to make the wasm target work.

I think something like https://github.com/rkusa/sqlite-vfs, but with no assumptions about the target environment (e.g. no usage of std stuff that is not available in the WASM environment) would be quite nice.

**Comment by rklaehn** [comment]
Here is a first draft of the traits:

https://github.com/rklaehn/sqlite-vfs/blob/wasm/src/lib.rs#L27

https://github.com/rklaehn/sqlite-vfs/blob/wasm/src/lib.rs#L43

Compared to the original, these allow you to specify how to generate randomness and current time. This is important for being able to support wasm.

**Comment by jalcine** [comment]
I've been very interested in doing this to add support for it running over something like [https://docs.rs/vfs/latest/vfs/](https://docs.rs/vfs/latest/vfs/). I was testing out implementing a VFS using [https://github.com/jalcine/sqlite-hypercore/](https://github.com/jalcine/sqlite-hypercore/) since <code>rusqlite</code> forwards <code>libsqlite3</code> so one could _attempt_ building a vfs. But having a nice wrapper would be a great add. I'd be happy to contribute from what I have to this repo.

(Originally published at: https://jacky.wtf/2022/5/007a)

**Comment by thomcc** [comment]
Yeah, I think this is actually a really important feature to add to rusqlite eventually (although it's going to be wildly unsafe). That said, it's worth noting that what SQLite considers a VFS is a bit more involved than what the `vfs` crate considers.

That said, it would probably not be no_std as you mention would be desirable in that link. Most of `rusqlite` uses std, and making that optional... I'm not sure. I guess it could reuse the VFS abstractions for some of them, if they existed, but it's tricky. Either way, it would be a lot of work, and I don't really have the time to do it. Already rusqlite is a project maintained in my spare time, and while I can't say if the same is true for gwenn, I suspect it is.

That said, for a VFS abstraction I have a few reasons I could justify it for work related things in the next year, so that's more plausible.

**Comment by jalcine** [comment]
I've [begun working on a PR](https://github.com/jalcine/rusqlite/pull/1) to first provide an inline set of traits, similar to what is available at https://github.com/rkusa/sqlite-vfs to rusqlite. Then I can work on a separate project that attempts to bridge https://docs.rs/vfs/latest/vfs/ generically. Like you mentioned, there's definitely some stuff that SQLite expects, and I'll see what I can do to work with it as safely as possible; most likely, I'd need to implement an additional trait to get those values and see if I can convince upstream to add those (and if not, then it's an auxiliary trait for people to implement to work with SQLite's VFS, which I think, is fair enough, it's an extension). This is personally critical to some stuff I have going on (but also it's a side project, so this is also a learning experience for me). I can spend about an hour or so each day on this, since it's _that_ tied up with the things I'm working on.

(Originally published at: https://jacky.wtf/2022/7/c3Wq)

**Comment by thomcc** [comment]
Okay. It's a very good feature, but they're pretty hard to write correctly (aside ones that just forward to other SQLite VFS's), both in terms of memory safety[^2] and avoiding database corruption[^1]

In general, I think support for custom SQLite VFS's is something that we'd merge, but probably not integration with the `vfs` crate. You should just publish your own crate that integrates the two if you need it, though.

If it would help, I'm reasonably familiar with SQLite's VFS layer and am usually around in the rusqlite discord (link in README), and can answer questions if you have any about safety and SQLite's expectations the like. It might be worth doing this to make sure the API is okay before you implement too much, to avoid needing large changes, but it's your call.

[^1]: For example, the `vfs` crate you link doesn't have any notion of file synchronization or locking, both of which which are usually needed to avoid corruption except in read-only cases.

[^2]: Despite the name, it's basically SQLite's entire OS abstraction, not just the filesystem.

**Comment by danthegoodman1** [comment]
this is dead in the water? Would love to implement a VFS on top of this

**Comment by jalcine** [comment]
@danthegoodman1 If you're brave enough, you can pick up where I left off with https://github.com/jalcine/rusqlite/pull/1. It's still in rough shape but with enough logging to make it viable. It's not an easy task and I'm not the best with unsafe code (currently!) so beware.

**Comment by danthegoodman1** [comment]
Unfortunately my rust is likely an order of magnitude too weak to be helpful there :/
